### PR TITLE
MAGN-8492 Small curve causes Dynamo to crash when showing background 3D preview in Revit. 2016

### PR DESCRIPTION
**Pull request by ikeough**
### Purpose

This PR addresses [MAGN-8492](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-8492). Revit cannot draw curves smaller than 1/64". We add a check whether the curve to be drawn is shorter than `Application.ShortCurveTolerance`. If so, we skip it.

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files

### Reviewers

@mjkkirschner 

### Cherry Picks:
- [ ] Revit2015
- [ ] Revit2017

### CODE REVIEW

File: test/Libraries/RevitIntegrationTests/RevitWatch3DViewModelTests.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+namespace RevitSystemTests
+{
+    [TestFixture]
+    public class RevitWatch3DViewModelTests : RevitSystemTestBase
+    {
+        [SetUp]
+        public void Setup()
+        {
+            var vm3D = ViewModel.Watch3DViewModels.FirstOrDefault(vm => vm is RevitWatch3DViewModel);
+            vm3D.Active = false;
```
Comment (Randy-Ma) [inline test/Libraries/RevitIntegrationTests/RevitWatch3DViewModelTests.cs]: I am curious why this is set again as it has been set in StartDynamo.

File: test/Libraries/RevitIntegrationTests/RevitWatch3DViewModelTests.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        }
+
+        [Test, TestModel(@".\empty.rfa")]
+        public void SmallCurve_Draw_DoesNotThrowException()
+        {
+            var vm3D = ViewModel.Watch3DViewModels.FirstOrDefault(vm => vm is RevitWatch3DViewModel);
+            Assert.NotNull(vm3D);
```
Comment (Randy-Ma) [inline test/Libraries/RevitIntegrationTests/RevitWatch3DViewModelTests.cs]: Shall we set vm3D.Active = true for this test case?

File: src/DynamoRevit/ViewModel/RevitWatch3DViewModel.cs [hunk 3]
```diff
@@ -221,7 +223,10 @@ private IEnumerable<GeometryObject> Tessellate(Curve curve)
             {
                 var xyz0 = new XYZ(verts[i] * conv, verts[i + 1] * conv, verts[i + 2] * conv);
                 var xyz1 = new XYZ(verts[i + 3] * conv, verts[i + 4] * conv, verts[i + 5] * conv);
-
+                if (xyz0.DistanceTo(xyz1) < DocumentManager.Instance.CurrentUIApplication.Application.ShortCurveTolerance)
+                {
+                    continue;
+                }
```
Comment (Randy-Ma) [inline src/DynamoRevit/ViewModel/RevitWatch3DViewModel.cs]: If a curve of 1' in length is tessellated into 1000 segments, then the curve won't be visible.

Shall we check the length of the curve, if it is too small then it won't even be faceted? If we do facet it, shall we use Polyline.Create() to create a polyline based on the facted points?

**Comment by Randy-Ma** [comment]
@ikeough 
I am seeing you are addressing an issue which I am also trying to fix. So I have put few comments.

File: test/Libraries/RevitIntegrationTests/RevitWatch3DViewModelTests.cs (hunk 2 above)
Comment (ikeough) [inline test/Libraries/RevitIntegrationTests/RevitWatch3DViewModelTests.cs]: You are correct. This was mistakenly set to false.

File: src/DynamoRevit/ViewModel/RevitWatch3DViewModel.cs (hunk 3 above)
Comment (ikeough) [inline src/DynamoRevit/ViewModel/RevitWatch3DViewModel.cs]: Yes, you are correct. The line will disappear. I considered this and figured that this was an acceptable loss. We are talking about lines that are shorter than 0.0013 units in Revit. `Polyline.Create` is a good idea, but I wonder if it is subject to the same limitations. That is, what if someone passes in a tiny little circle?

Comment (mjkkirschner) [inline src/DynamoRevit/ViewModel/RevitWatch3DViewModel.cs]: If Revit Polylines can have segments shorter than the tolerance, then discarding initial lines below that tolerance should avoid that problem... correct?

Comment (Randy-Ma) [inline src/DynamoRevit/ViewModel/RevitWatch3DViewModel.cs]: Yes, @mjkkirschner has got what I have suggested.

Just to make it more clear:
1). Check the curve's length at the very beginning of the function, if it is less than the tolerance, return from there;
2). Facet the curve;
3). Create one polyline from the faceted points;

File: src/DynamoRevit/ViewModel/DynamoRevitViewModel.cs [hunk 4]
```diff
@@ -1,5 +1,5 @@
 ﻿using System;
-
+using System.Drawing;
```
Comment (mjkkirschner) [inline src/DynamoRevit/ViewModel/DynamoRevitViewModel.cs]: do we require system drawing for some reason here?

File: test/Libraries/RevitTestServices/RevitSystemTestBase.cs [hunk 5]
```diff
@@ -221,12 +222,18 @@ protected override void StartDynamo(TestSessionConfiguration testConfig)
 
                 Model = DynamoRevit.RevitDynamoModel;
 
-                this.ViewModel = DynamoViewModel.Start(
+                this.ViewModel = DynamoRevitViewModel.Start(
```
Comment (mjkkirschner) [inline test/Libraries/RevitTestServices/RevitSystemTestBase.cs]: was this a previous mistake somewhere or has something changed fundamentally?

**Comment by mjkkirschner** [comment]
@ikeough I will leave the decision up to you if we should create a more robust, but complex algorithm for short curve rejection in this pr, otherwise I just have one small comment and a question.

File: src/DynamoRevit/ViewModel/RevitWatch3DViewModel.cs [hunk 6]
```diff
@@ -210,20 +212,18 @@ private IEnumerable<GeometryObject> Tessellate(Curve curve)
...
+            var scaledXYZs = new List<XYZ>();
+            for (var i = 0; i < verts.Count; i+= 3)
             {
-                var xyz0 = new XYZ(verts[i] * conv, verts[i + 1] * conv, verts[i + 2] * conv);
-                var xyz1 = new XYZ(verts[i + 3] * conv, verts[i + 4] * conv, verts[i + 5] * conv);
-
-                result.Add(Line.CreateBound(xyz0, xyz1));
+                scaledXYZs.Add(new XYZ(verts[i] * conv, verts[i + 1] * conv, verts[i + 2] * conv));
             }
+            result.Add(PolyLine.Create(scaledXYZs));
```
Comment (ikeough) [inline src/DynamoRevit/ViewModel/RevitWatch3DViewModel.cs]: @Randy-Ma @mjkkirschner As it turns out, `Polyline.Create` is incredibly tolerant of small lines. It does not seem subject, at least when created lines for the background preview, to the short curve tolerance normally applied when drawing other types of curves in Revit. For this reason, I've switched this entire method of drawing curves, including straight curves, over to using that method.

File: test/Libraries/RevitTestServices/RevitSystemTestBase.cs (hunk 5 above)
Comment (ikeough) [inline test/Libraries/RevitTestServices/RevitSystemTestBase.cs]: That's a good question. I am not sure why this was done originally, but it seemed that we should be testing the way that a user would be running, using `DynamoRevitViewModel`. It is possible that when we run this through testing, the reason for this change will become obvious.

File: src/DynamoRevit/ViewModel/RevitWatch3DViewModel.cs [hunk 7]
```diff
@@ -104,6 +104,8 @@ protected override void OnNodePropertyChanged(object sender, PropertyChangedEven
 
         private void Draw()
         {
+            if (!Active) return;
```
Comment (mjkkirschner) [inline src/DynamoRevit/ViewModel/RevitWatch3DViewModel.cs]: nice

File: src/DynamoRevit/ViewModel/RevitWatch3DViewModel.cs (hunk 6 above)
Comment (mjkkirschner) [inline src/DynamoRevit/ViewModel/RevitWatch3DViewModel.cs]: @ikeough cool - so there is no need to check the length of the polyline?

**Comment by mjkkirschner** [comment]
@ikeough this looks good to merge pending one question

**Pull request merged**